File: src/sense_hat/yaw_rotation_imu_mock.py
from collections.abc import Callable

from builtin_interfaces.msg import Time
from geometry_msgs.msg import Vector3
from sensor_msgs.msg import Imu

from sense_hat.imu_message import G, new_imu_message
from sense_hat.imu_strategy_protocol import ImuStrategy


def _default_stamp() -> Time:
    return Time()


class YawRotationImuMock(ImuStrategy):
    """A mock IMU that rotates around the yaw axis."""

    def __init__(
        self,
        frame_id: str = "",
        stamp_factory: Callable[[], Time] | None = None,
        angular_velocity_z: float = 0.25,
        g: float = G,
    ) -> None:
        """Initialize the mock IMU."""
        self.frame_id = frame_id
        self.stamp_factory = stamp_factory if stamp_factory is not None else _default_stamp
        self.angular_velocity_z = angular_velocity_z
        self.g = g

    def get_imu_message(self) -> Imu:
        """Get the IMU message."""
        msg = new_imu_message(self.stamp_factory(), self.frame_id)
        msg.angular_velocity = Vector3(x=0.0, y=0.0, z=self.angular_velocity_z)
        msg.linear_acceleration = Vector3(x=0.0, y=0.0, z=self.g)
        return msg

    def close(self) -> None:
        """Release strategy resources."""

File: src/sense_hat_bridge/src/yaw_rotation_imu_mock.cpp
#include "sense_hat_bridge/yaw_rotation_imu_mock.hpp"

#include <algorithm>
#include <cmath>
#include <utility>

#include "sense_hat_bridge/imu_message.hpp"

namespace sense_hat_bridge {

namespace {

builtin_interfaces::msg::Time default_stamp() {
  return builtin_interfaces::msg::Time{};
}

}  // namespace

YawRotationImuMock::YawRotationImuMock(std::string frame_id, StampFactory stamp_factory,
                                       YawRotationRampParams yaw_rotation_ramp_params)
    : frame_id_(std::move(frame_id)),
      stamp_factory_(stamp_factory ? std::move(stamp_factory) : StampFactory{default_stamp}),
      yaw_rotation_ramp_params_(yaw_rotation_ramp_params),
      current_angular_velocity_z_(yaw_rotation_ramp_params.start_angular_velocity_z) {}

std::optional<sensor_msgs::msg::Imu> YawRotationImuMock::get_imu_message() {
  auto msg = new_imu_message(stamp_factory_(), frame_id_);
  const double progress = std::clamp(
      static_cast<double>(ramp_step_) / (yaw_rotation_ramp_params_.duration_s * kMockRateHz), 0.0,
      1.0);
  const double shape = 4.0;
  const double alpha =
      progress >= 1.0 ? 1.0 : (1.0 - std::exp(-shape * progress)) / (1.0 - std::exp(-shape));
  const double angular_velocity_z = yaw_rotation_ramp_params_.start_angular_velocity_z +
                                    ((yaw_rotation_ramp_params_.end_angular_velocity_z -
                                      yaw_rotation_ramp_params_.start_angular_velocity_z) *
                                     alpha);
  current_angular_velocity_z_ = angular_velocity_z;
  msg.angular_velocity.z = angular_velocity_z;
  msg.angular_velocity.x = 0.0;
  msg.angular_velocity.y = 0.0;
  msg.linear_acceleration.x = 0.0;
  msg.linear_acceleration.y = 0.0;
  msg.linear_acceleration.z = kGravityMetersPerSecondSquared;
  ramp_step_++;
  return msg;
}

}  // namespace sense_hat_bridge
